starlette/schemas.py

```diff
@@ -21,42 +21,66 @@ def render(self, content: typing.Any) -> bytes:
         return yaml.dump(content, default_flow_style=False).encode("utf-8")
 
 
+class Endpoint(typing.NamedTuple):
+    path: str
+    http_method: str
+    func: typing.Callable
+
+
 class BaseSchemaGenerator:
     def get_schema(self, routes: typing.List[BaseRoute]) -> dict:
         raise NotImplementedError()  # pragma: no cover
 
+    def get_endpoints(
+        self, routes: typing.List[BaseRoute]
+    ) -> typing.Generator[Endpoint, None, None]:
+        """
+        Given the routes, yields the following information:
 
-class SchemaGenerator(BaseSchemaGenerator):
-    def __init__(self, base_schema: dict) -> None:
-        assert yaml is not None, "`pyyaml` must be installed to use SchemaGenerator."
-        self.base_schema = base_schema
-
-    def get_schema(self, routes: typing.List[BaseRoute]) -> dict:
-        paths = {}  # type: dict
-
+        - path
+            eg: /users/
+        - http_method
+            one of 'get', 'post', 'put', 'patch', 'delete', 'options'
+        - func
+            method ready to extract the docstring
+        """
         for route in routes:
             if not isinstance(route, Route) or not route.include_in_schema:
                 continue
 
             if inspect.isfunction(route.endpoint) or inspect.ismethod(route.endpoint):
-                docstring = route.endpoint.__doc__
                 for method in route.methods or ["GET"]:
                     if method == "HEAD":
                         continue
-                    if route.path not in paths:
-                        paths[route.path] = {}
-                    data = yaml.safe_load(docstring) if docstring else {}
-                    paths[route.path][method.lower()] = data
+                    yield Endpoint(route.path, method.lower(), route.endpoint)
```

Comment:
Let's just return a list of `Endpoint` instead.

Reply:
okay, I'll update it. It feels a bit weird iterating once to collect the endpoints info and then iterating again to generate the schema, do you find it less readable with the `yield`? or am I adding too much unnecessary complexity?

Reply:
I leave it with the array for now